src/core/Ekubo.sol

```diff
@@ -0,0 +1,486 @@
+                data.length := sub(data.length, 0x02)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            data = Decoder.updateState(state, notes, data);
+            int256 amountSpecified;
+            unchecked {
+                amountSpecified = int256((state.sell.amount * bps).unsafeDiv(BASIS));
+            }
+            bool isToken1;
+            {
+                (IERC20 sellToken, IERC20 buyToken) = (state.sell.token, state.buy.token);
+                assembly ("memory-safe") {
+                    sellToken := and(_ADDRESS_MASK, sellToken)
+                    buyToken := and(_ADDRESS_MASK, buyToken)
+                    isToken1 :=
+                        or(
+                            eq(buyToken, 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee),
+                            and(iszero(eq(sellToken, 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)), lt(buyToken, sellToken))
+                        )
+                }
+                (poolKey.token0, poolKey.token1) = isToken1.maybeSwap(address(sellToken), address(buyToken));
+            }
+
+            {
+                bytes32 config;
+                assembly ("memory-safe") {
+                    config := calldataload(data.offset)
+                    data.offset := add(0x20, data.offset)
+                    data.length := sub(data.length, 0x20)
+                    // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+                }
+                poolKey.config = Config.wrap(config);
+            }
+
+            uint256 skipAhead;
+            assembly ("memory-safe") {
+                skipAhead := calldataload(data.offset)
+                data.offset := add(0x20, data.offset)
+                data.length := sub(data.length, 0x20)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            Decoder.overflowCheck(data);
+
+            {
+                SqrtRatio sqrtRatio = SqrtRatio.wrap(
+                    uint96(isToken1.ternary(uint256(79227682466138141934206691491), uint256(4611797791050542631)))
+                );
+                (int256 delta0, int256 delta1) = IEkuboCore(msg.sender).unsafeSwap(
+                    poolKey, int128(amountSpecified), isToken1, sqrtRatio, skipAhead
+                );
+                // Ekubo sign convention here is backwards compared to UniV4/BalV3/PancakeInfinity
+                // settledSellAmount is positive, settledBuyAmount is negative
+                (int256 settledSellAmount, int256 settledBuyAmount) = isToken1.maybeSwap(delta0, delta1);
+
+                state.sell.amount -= settledSellAmount.asCredit(state.sell.token);
+
+                // If `state.buy.amount()` overflows an `int128`, we'll get a revert inside the
+                // vault later. We cannot overflow a `uint256`.
```

Comment:
Is this true? What will happen if we overflow an `int128`? I wonder if it might be better to leave the argument to `unsafeSwap` as `int256` (omitting the `SIGNEXTEND` before writing the value to memory) and then let the ABI decoding in `CORE` serve as our guard against overflow.